diff --git a/Makefile.am b/Makefile.am
index 445c056..c212884 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -71,7 +71,7 @@ libbsdiff_la_SOURCES = \
 	src/sufsort.c
 
 libbsdiff_la_LIBADD = \
-	$(zlib_LIBS)
+	$(zlib_LIBS) -lzopfli
 
 if ENABLE_LZMA
 libbsdiff_la_LIBADD += \
diff --git a/src/diff.c b/src/diff.c
index 85ac86b..4a8e9c8 100644
--- a/src/diff.c
+++ b/src/diff.c
@@ -57,6 +57,7 @@ __FBSDID
 #include <pthread.h>
 #include <assert.h>
 #include <sys/mman.h>
+#include <zopfli.h>
 
 #include "bsheader.h"
 
@@ -120,6 +121,29 @@ static inline void offtout(int64_t x, u_char *buf)
 	*((int64_t *)buf) = htole64(x);
 }
 
+static int compress2gzip_zopfli(Bytef *dest, size_t *destLen,
+			 const Bytef *source, uLong sourceLen)
+{
+	ZopfliOptions options;
+	unsigned char* out = 0;
+	size_t outsize = 0;
+
+	ZopfliInitOptions(&options);
+	options.numiterations = 15;	
+
+	ZopfliCompress(&options, ZOPFLI_FORMAT_GZIP, source, sourceLen, &out, &outsize);
+
+	if (outsize < *destLen) {
+		memcpy(dest, out, outsize);
+		*destLen = outsize;
+		free(out);
+		return Z_OK;
+	}
+	free(out);
+	return Z_STREAM_ERROR;
+
+}
+
 /* zlib provides compress2, which deflates to deflate (zlib) format. This is
  * unfortunately distinct from gzip format in that the headers wrapping the
  * decompressed data are different. gbspatch reads gzip-compressed data using
@@ -135,6 +159,10 @@ static int compress2gzip(Bytef *dest, size_t *destLen,
 	z_stream stream;
 	int err;
 
+	/* for small-ish blocks we use zopfli */
+	if (sourceLen < 128 * 1024)
+		return compress2gzip_zopfli(dest, destLen, source, sourceLen);
+
 	stream.next_in = (Bytef *)source;
 	stream.avail_in = (uInt)sourceLen;
 
